docs/src/configuration.md
# Configuration

When using the optimiser, you change it's behaviour via a configuration file specified as a TOML.
The configuration file serves two purposes.
Firstly, it makes it easier for you to track various settings and their impacts.
Secondly, if something breaks, it makes it easier for us to reproduce what went wrong.

An example configuration TOML file might look as below.

``` toml
id = "0x6f8a032b4b1ee622ef2f0fc091bdbb98cfae81a3"
writedir = "data"
max_allocations = 10
whitelist = []
blacklist = []
frozenlist = []
pinnedlist = []
allocation_lifetime = 28
gas = 100
min_signal = 100
verbose = true
num_reported_options = 2
execution_mode = "none"
opt_mode = "optimal"
network_subgraph_endpoint = "https://gateway.thegraph.com/api/{api-key}/subgraphs/id/DZz4kDTdmzWLWsV373w2bSmoar3umKKH9y82SUKr5qmp"
protocol_network = "arbitrum"
syncing_networks = ["mainnet", "gnosis", "arbitrum-one", "arbitrum"]
```

### Detailed Field Descriptions

- `id::String`: The ID of the indexer for whom we're optimising. No default value.
- `network_subgraph_endpoint::String`: The network subgraph endpoint to query. The optimizer
    support any network (such as mainnet, goerli, arbitrum-one, arbitrum-goerli) as long as the
    provided API serves the query requests. If unspecified,
    `"https://api.thegraph.com/subgraphs/name/graphprotocol/graph-network-mainnet"`
- `writedir::String`: The directory to which to write the results of optimisation.
    If don't specify `readdir`, `writedir` also specifies the path to which to save
    the input data tables. If unspecified, `"."`
- `readdir::Union{String, Nothing}`: The directory from which to read saved data tables.
    This speeds up the process as we won't have to query the network subgraph for the
    relevant data. If you don't specify `readdir`, we will query your specified
    `network_subgraph_endpoint` for the data and write it to CSV files in `writedir`.
    This way, you can use your previous `writedir` as your `readdir` in future runs.
    If unspecified, `nothing`
- `whitelist::Vector{String}`: A list of subgraph IPFS hashes that you want to consider
    as candidates to which to allocate. If you leave this empty, we'll assume all subgraphs
    are in the whitelist. If unspecified, `String[]`
- `blacklist::Vector{String}`: A list of subgraph IPFS hashes that you do not want to
    consider allocating to. For example, this list could include broken subgraphs or
    subgraphs that you don't want to index. If unspecified, `String[]`
- `frozenlist::Vector{String}`: If you have open allocations that you don't want to change,
    add the corresponding subgraph IPFS hashes to this list. If unspecified, `String[]`
- `pinnedlist::Vector{String}`: If you have subgraphs that you absolutely want to be
    allocated to, even if only with a negligible amount of GRT, add it to this list.
    If unspecified, `String[]`
- `allocation_lifetime::Integer`: The number of epochs for which you expect the allocations
    the optimiser finds to be open. If unspecified, `28`
- `gas::Real`: The estimated gas cost in GRT to open/close allocations. If unspecified, `100`
- `min_signal::Real`: The minimum amount of signal in GRT that must be on a subgraph
    in order for you to consider allocating to it. If unspecified, `100`
- `max_allocations::Integer`: The maximum number of new allocations you'd like the optimiser
    to consider opening. If unspecified, `10`
- `num_reported_options::Integer`: The number of proposed allocation strategies to report.
    For example, if you select `10` we'd report best 10 allocation strategies ranked by
    profit. Options are reported to a *report.json* in your `writedir`. If unspecified, `1`
- `verbose::Bool`: If true, the optimiser will print details about what it is doing to
    stdout. If unspecified, `false`
- `execution_mode::String`: How the optimiser should execute the allocation strategies it
    finds. Options are `"none"`, which won't do anything, `"actionqueue"`, which will
    push actions to the action queue, and `"rules"`, which will generate indexing rules.
    If unspecified, `"none"`
- `indexer_url::Union{String, Nothing}`: The URL of the indexer management server you want
    to execute the allocation strategies on. If you specify `"actionqueue"`, you must also
    specify `indexer_url`. If unspecified, `nothing`
- `opt_mode::String`: We support three optimisation modes. One is `"fastnogas"`. This mode does
    not consider gas costs and optimises allocation amount over all subgraph deployments.
    Second one is `"fastgas"`. This mode is fast, but may not find the optimal strategy and
    could potentially fail to converge. This mode is also used to the top
    `num_reported_options` allocation strategies. The final mode is `"optimal"`.
    This mode is slower, but it satisfies stronger optimality conditions.
    It will find strategies at least as good as `"fast"`, but not guaranteed to be better.
    By default, `"optimal"`
- `protocol_network::String`: Defines the protocol network that allocation transactions
    should be sent to. The current protocol network options are "mainnet", "goerli",
    "arbitrum", and "arbitrum-goerli". By default, `"mainnet"`
- `syncing_networks::Vector{String}`: The list of syncing networks to support when selecting
    the set of possible subgraphs. This list should match the networks available to your
    graph-node. By default, the list is a singleton of your protocol network

### Example Configurations

#### ActionQueue

Set `execution_mode` to `"actionqueue"` and provide an `indexer_url`.

``` toml
id = "0xd75c4dbcb215a6cf9097cfbcc70aab2596b96a9c"
writedir = "data"
readdir = "data"
max_allocations = 10
whitelist = []
blacklist = []
frozenlist = []
pinnedlist = []
allocation_lifetime = 28
gas = 100
min_signal = 100
verbose = true
num_reported_options = 2
execution_mode = "actionqueue"
indexer_url = "https://localhost:8000"
```

#### Indexer Rules

Change `execution_mode` to `"rules"`.

``` toml
id = "0xd75c4dbcb215a6cf9097cfbcc70aab2596b96a9c"
writedir = "data"
readdir = "data"
max_allocations = 10
whitelist = []
blacklist = []
frozenlist = []
pinnedlist = []
allocation_lifetime = 28
gas = 100
min_signal = 100
verbose = true
num_reported_options = 2
execution_mode = "rules"
```

#### Query data Instead of Reading Local CSVs

Just don't specify the `readdir`.

``` toml
id = "0xd75c4dbcb215a6cf9097cfbcc70aab2596b96a9c"
writedir = "data"
max_allocations = 10
whitelist = []
blacklist = []
frozenlist = []
pinnedlist = []
allocation_lifetime = 28
gas = 100
min_signal = 100
verbose = true
num_reported_options = 2
execution_mode = "none"
network_subgraph_endpoint = "https://gateway.thegraph.com/api/{api-key}/subgraphs/id/DZz4kDTdmzWLWsV373w2bSmoar3umKKH9y82SUKr5qmp"
protocol_network = "arbitrum"
syncing_networks = ["mainnet", "gnosis", "arbitrum-one", "arbitrum"]
```

#### Quiet Mode

We set `verbose` to `false` here to surpress info messages.

``` toml
id = "0xd75c4dbcb215a6cf9097cfbcc70aab2596b96a9c"
writedir = "data"
readdir = "data"
max_allocations = 10
whitelist = []
blacklist = []An example configuration TOML file might look as below.

``` toml
id = "0x6f8a032b4b1ee622ef2f0fc091bdbb98cfae81a3"
writedir = "data"
max_allocations = 10
whitelist = []
blacklist = []
frozenlist = []
pinnedlist = []
allocation_lifetime = 28
gas = 100
min_signal = 100
verbose = true
num_reported_options = 2
execution_mode = "none"
opt_mode = "optimal"
network_subgraph_endpoint = "https://gateway.thegraph.com/api/{api-key}/subgraphs/id/DZz4kDTdmzWLWsV373w2bSmoar3umKKH9y82SUKr5qmp"
protocol_network = "arbitrum"
syncing_networks = ["mainnet", "gnosis", "arbitrum-one", "arbitrum"]
```


### Detailed Field Descriptions

- `id::String`: The ID of the indexer for whom we're optimising. No default value.
- `network_subgraph_endpoint::String`: The network subgraph endpoint to query. The optimizer
    support any network (such as mainnet, goerli, arbitrum-one, arbitrum-goerli) as long as the
    provided API serves the query requests. If unspecified,
    `"https://api.thegraph.com/subgraphs/name/graphprotocol/graph-network-mainnet"`
- `writedir::String`: The directory to which to write the results of optimisation.
    If don't specify `readdir`, `writedir` also specifies the path to which to save
    the input data tables. If unspecified, `"."`
- `readdir::Union{String, Nothing}`: The directory from which to read saved data tables.
    This speeds up the process as we won't have to query the network subgraph for the
    relevant data. If you don't specify `readdir`, we will query your specified
    `network_subgraph_endpoint` for the data and write it to CSV files in `writedir`.
    This way, you can use your previous `writedir` as your `readdir` in future runs.
    If unspecified, `nothing`
- `whitelist::Vector{String}`: A list of subgraph IPFS hashes that you want to consider
    as candidates to which to allocate. If you leave this empty, we'll assume all subgraphs
    are in the whitelist. If unspecified, `String[]`
- `blacklist::Vector{String}`: A list of subgraph IPFS hashes that you do not want to
    consider allocating to. For example, this list could include broken subgraphs or
    subgraphs that you don't want to index. If unspecified, `String[]`
- `frozenlist::Vector{String}`: If you have open allocations that you don't want to change,
    add the corresponding subgraph IPFS hashes to this list. If unspecified, `String[]`
- `pinnedlist::Vector{String}`: If you have subgraphs that you absolutely want to be
    allocated to, even if only with a negligible amount of GRT, add it to this list.
    If unspecified, `String[]`
- `allocation_lifetime::Integer`: The number of epochs for which you expect the allocations
    the optimiser finds to be open. If unspecified, `28`
- `gas::Real`: The estimated gas cost in GRT to open/close allocations. If unspecified, `100`
- `min_signal::Real`: The minimum amount of signal in GRT that must be on a subgraph
    in order for you to consider allocating to it. If unspecified, `100`
- `max_allocations::Integer`: The maximum number of new allocations you'd like the optimiser
    to consider opening. If unspecified, `10`
- `num_reported_options::Integer`: The number of proposed allocation strategies to report.
    For example, if you select `10` we'd report best 10 allocation strategies ranked by
    profit. Options are reported to a *report.json* in your `writedir`. If unspecified, `1`
- `verbose::Bool`: If true, the optimiser will print details about what it is doing to
    stdout. If unspecified, `false`
- `execution_mode::String`: How the optimiser should execute the allocation strategies it
    finds. Options are `"none"`, which won't do anything, `"actionqueue"`, which will
    push actions to the action queue, and `"rules"`, which will generate indexing rules.
    If unspecified, `"none"`
- `indexer_url::Union{String, Nothing}`: The URL of the indexer management server you want
    to execute the allocation strategies on. If you specify `"actionqueue"`, you must also
    specify `indexer_url`. If unspecified, `nothing`
- `opt_mode::String`: We support three optimisation modes. One is `"fastnogas"`. This mode does
    not consider gas costs and optimises allocation amount over all subgraph deployments.
    Second one is `"fastgas"`. This mode is fast, but may not find the optimal strategy and
    could potentially fail to converge. This mode is also used to the top
    `num_reported_options` allocation strategies. The final mode is `"optimal"`.
    This mode is slower, but it satisfies stronger optimality conditions.
    It will find strategies at least as good as `"fast"`, but not guaranteed to be better.
    By default, `"optimal"`
- `protocol_network::String`: Defines the protocol network that allocation transactions
    should be sent to. The current protocol network options are "mainnet", "goerli",
    "arbitrum", and "arbitrum-goerli". By default, `"mainnet"`
- `syncing_networks::Vector{String}`: The list of syncing networks to support when selecting
    the set of possible subgraphs. This list should match the networks available to your
    graph-node. By default, the list is a singleton of your protocol network

### Example Configurations

#### ActionQueue

Set `execution_mode` to `"actionqueue"` and provide an `indexer_url`.

``` toml
id = "0xd75c4dbcb215a6cf9097cfbcc70aab2596b96a9c"
writedir = "data"
readdir = "data"
max_allocations = 10
whitelist = []
blacklist = []
frozenlist = []
pinnedlist = []
allocation_lifetime = 28
gas = 100
min_signal = 100
verbose = true
num_reported_options = 2
execution_mode = "actionqueue"
indexer_url = "https://localhost:8000"
protocol_network = "arbitrum"
syncing_network = ["mainnet"]
```

#### Indexer Rules

Change `execution_mode` to `"rules"`.

``` toml
id = "0xd75c4dbcb215a6cf9097cfbcc70aab2596b96a9c"
writedir = "data"
readdir = "data"
max_allocations = 10
whitelist = []
blacklist = []
frozenlist = []
pinnedlist = []
allocation_lifetime = 28
gas = 100
min_signal = 100
verbose = true
num_reported_options = 2
execution_mode = "rules"
```

#### Query data Instead of Reading Local CSVs

Just don't specify the `readdir`.

``` toml
id = "0xd75c4dbcb215a6cf9097cfbcc70aab2596b96a9c"
writedir = "data"
max_allocations = 10
whitelist = []
blacklist = []
frozenlist = []
pinnedlist = []
allocation_lifetime = 28
gas = 100
min_signal = 100
verbose = true
num_reported_options = 2
execution_mode = "none"
```

#### Query data for specified networks

Specify the network subgraph endpoint for networks other than The Graph network on Ethereum mainnet. Here we use the endpoint to Ethereum Sepolia network subgraph.

``` toml
id = "0xE9a1CABd57700B17945Fd81feeFba82340D9568F"
network_subgraph_endpoint = "https://gateway.thegraph.com/api/{api-key}/subgraphs/id/8pVKDwHniAz87CHEQsiz2wgFXGZXrbMDkrxgauVVfMJC"
```

Other available endpoints examples are
- Mainnet (default): https://gateway.thegraph.com/api/{api-key}/subgraphs/id/9Co7EQe5PgW3ugCUJrJgRv4u9zdEuDJf8NvMWftNsBH8
- Arbitrum-One: https://gateway.thegraph.com/api/{api-key}/subgraphs/id/DZz4kDTdmzWLWsV373w2bSmoar3umKKH9y82SUKr5qmp
- Arbitrum-Sepolia: https://gateway.thegraph.com/api/{api-key}/subgraphs/id/3xQHhMudr1oh69ut36G2mbzpYmYxwqCeU6wwqyCDCnqV

#### Quiet Mode

We set `verbose` to `false` here to surpress info messages.

``` toml
id = "0xd75c4dbcb215a6cf9097cfbcc70aab2596b96a9c"
writedir = "data"
readdir = "data"
max_allocations = 10
whitelist = []
blacklist = []
frozenlist = []
pinnedlist = []
allocation_lifetime = 28
gas = 100
min_signal = 100
verbose = false
num_reported_options = 2
execution_mode = "none"
```

#### Whitelisting Subgraphs

Add some subgraph deployment IDs to the `whitelist`.
If, in addition or instead you want to use `blacklist`, `frozenlist`, or `pinnedlist`, you can
similarly add subgraph deployment IDs to those lists.
Notice that we did not change `max_allocations` here.
If `max_allocations` exceeds the number of available subgraphs (2 in this case), the code will
treat the number of available subgraphs as `max_allocations`.

``` toml
id = "0xd75c4dbcb215a6cf9097cfbcc70aab2596b96a9c"
writedir = "data"
readdir = "data"
max_allocations = 10
whitelist = [
    "QmUVskWrz1ZiQZ76AtyhcfFDEH1ELnRpoyEhVL8p6NFTbR",
    "QmcBSr5R3K2M5tk8qeHFaX8pxAhdViYhcKD8ZegYuTcUhC"
]
blacklist = []
frozenlist = []
pinnedlist = []
allocation_lifetime = 28
gas = 100
min_signal = 100
verbose = false
num_reported_options = 2
execution_mode = "none"
```


2
execution_mode = "none"
```

#### Whitelisting Subgraphs

Add some subgraph deployment IDs to the `whitelist`.
If, in addition or instead you want to use `blacklist`, `frozenlist`, or `pinnedlist`, you can
similarly add subgraph deployment IDs to those lists.
Notice that we did not change `max_allocations` here.
If `max_allocations` exceeds the number of available subgraphs (2 in this case), the code will
treat the number of available subgraphs as `max_allocations`.

``` toml
id = "0xd75c4dbcb215a6cf9097cfbcc70aab2596b96a9c"
writedir = "data"
readdir = "data"
max_allocations = 10
whitelist = [
    "QmUVskWrz1ZiQZ76AtyhcfFDEH1ELnRpoyEhVL8p6NFTbR",
    "QmcBSr5R3K2M5tk8qeHFaX8pxAhdViYhcKD8ZegYuTcUhC"
]
blacklist = []
frozenlist = []
pinnedlist = []
allocation_lifetime = 28
gas = 100
min_signal = 100
verbose = false
num_reported_options = 2
execution_mode = "none"
```


CHANGELOG.md
# Changelog

# [2.7.0](https://github.com/graphprotocol/allocation-optimizer/compare/v2.6.0...v2.7.0) (2024-06-18)


### Features

* add fastnogas optimisation option ([7cde58f](https://github.com/graphprotocol/allocation-optimizer/commit/7cde58f581658421cb83dc4d53a2f53657ab0212))

# [2.6.0](https://github.com/graphprotocol/allocation-optimizer/compare/v2.5.1...v2.6.0) (2024-06-18)


### Features

* upgrade to use julia 1.10 ([d71cb11](https://github.com/graphprotocol/allocation-optimizer/commit/d71cb11133c328720ac0bb7a2cecfd8175263f86))

## [2.5.1](https://github.com/graphprotocol/allocation-optimizer/compare/v2.5.0...v2.5.1) (2024-01-25)


### Bug Fixes

* network subgraph FDS schema update ([2556c59](https://github.com/graphprotocol/allocation-optimizer/commit/2556c59342ebc0df353a14d5366336fb4227b9f3))

# [2.5.0](https://github.com/graphprotocol/allocation-optimizer/compare/v2.4.2...v2.5.0) (2023-09-27)


### Features

* support protocol_network config to actionqueue ([bba8173](https://github.com/graphprotocol/allocation-optimizer/commit/bba81730ba9d5eaaa8c6e7eda9fba8ba4e0dbd4d))
* support syncing_network filter in subgraph selection ([09a3b38](https://github.com/graphprotocol/allocation-optimizer/commit/09a3b3861c77c6b0448f70a9473a856e7036b698))

## [2.4.2](https://github.com/graphprotocol/allocation-optimizer/compare/v2.4.1...v2.4.2) (2023-09-07)


### Bug Fixes

* allow no current allocations ([4438cf5](https://github.com/graphprotocol/allocation-optimizer/commit/4438cf56851d9a4e154a60013e12b89b128bd644))
* N/A options for some k less than max_allocations ([8e4f8a0](https://github.com/graphprotocol/allocation-optimizer/commit/8e4f8a0b87a67ac115879b04d5f63d0912c4ed79))

## [2.4.1](https://github.com/graphprotocol/allocation-optimizer/compare/v2.4.0...v2.4.1) (2023-09-05)


### Bug Fixes

* linear indexing reward for network token issuance ([85d5ac7](https://github.com/graphprotocol/allocation-optimizer/commit/85d5ac7c50e66b98192c69e5e5e2a195cdc6a7af))

# [2.4.0](https://github.com/graphprotocol/allocation-optimizer/compare/v2.3.1...v2.4.0) (2023-05-11)


### Bug Fixes

* PGO computes optimal value for each sparsity ([2121343](https://github.com/graphprotocol/allocation-optimizer/commit/212134307b652f6b83bbd9096cd152e4ac2b55b2))


### Features

* PGO stops when it finds the same value twice ([d96592d](https://github.com/graphprotocol/allocation-optimizer/commit/d96592d2ce1658e6aeedd4cc6c0f010ce234c5d2))


### Performance Improvements

* Disable bounds checking on opt loops ([8b0ae3c](https://github.com/graphprotocol/allocation-optimizer/commit/8b0ae3c461c31fe3fd75a0a0311fce877f381182))

## [2.3.1](https://github.com/graphprotocol/allocation-optimizer/compare/v2.3.0...v2.3.1) (2023-05-03)


### Bug Fixes

* Correcting K for deniedAt filtering ([5840fbd](https://github.com/graphprotocol/allocation-optimizer/commit/5840fbdf1e7b8d9dbd3783f8c0f61c2d7ce012a5))

# [2.3.0](https://github.com/graphprotocol/allocation-optimizer/compare/v2.2.1...v2.3.0) (2023-04-27)


### Bug Fixes

* Filter out subgraphs with nonzero deniedAt ([219d4e0](https://github.com/graphprotocol/allocation-optimizer/commit/219d4e0930705e28e241e4d2e71169a0838dbb5e)), closes [#37](https://github.com/graphprotocol/allocation-optimizer/issues/37)


### Features

* deniedAt field accessor ([a37263b](https://github.com/graphprotocol/allocation-optimizer/commit/a37263bd9d193684423de28d37eb2846a794167b))
* Function gets indices of rewarded subgraphs ([113a86e](https://github.com/graphprotocol/allocation-optimizer/commit/113a86e6aea54ac07330cb8b6175a1919c37a503))
* Query deniedAt ([a7acaa4](https://github.com/graphprotocol/allocation-optimizer/commit/a7acaa4d4d451378f22aac46334120608033e5dc))

## [2.2.1](https://github.com/graphprotocol/allocation-optimizer/compare/v2.2.0...v2.2.1) (2023-04-27)


### Bug Fixes

* Pinnedlist behaviour fixed ([9cc96ba](https://github.com/graphprotocol/allocation-optimizer/commit/9cc96baf85d1be9e890d91f4bb2067e095d99ce2)), closes [#38](https://github.com/graphprotocol/allocation-optimizer/issues/38)

# [2.2.0](https://github.com/graphprotocol/allocation-optimizer/compare/v2.1.2...v2.2.0) (2023-04-27)


### Features

* Descriptive error when bad `readdir` ([66b264a](https://github.com/graphprotocol/allocation-optimizer/commit/66b264a00c878d9f545ebe552d7cab6b2e365d56))
* Warn users about not filtering deniedAt ([630f6b5](https://github.com/graphprotocol/allocation-optimizer/commit/630f6b581bdc3bd23dec00cfa78689da717751b0))

## [2.1.2](https://github.com/graphprotocol/allocation-optimizer/compare/v2.1.1...v2.1.2) (2023-04-24)


### Bug Fixes

* Pinned subgraphs are now whitelisted ([ddacc55](https://github.com/graphprotocol/allocation-optimizer/commit/ddacc555df08730b3fce04e83ce2aa06dc7a3b7e))

## [2.1.1](https://github.com/graphprotocol/allocation-optimizer/compare/v2.1.0...v2.1.1) (2023-04-17)


### Bug Fixes

* RoundDown when converting from BigFloat ([f8cce14](https://github.com/graphprotocol/allocation-optimizer/commit/f8cce14c26b046b2aebeae681ad57e8ef08b6e5c))

# [2.1.0](https://github.com/graphprotocol/allocation-optimizer/compare/v2.0.1...v2.1.0) (2023-04-17)


### Bug Fixes

* Allow for views ([2476317](https://github.com/graphprotocol/allocation-optimizer/commit/2476317597c1de2073142c42c0b0a9915982a8f4))
* Minimisation needs to use negative profit ([470a72e](https://github.com/graphprotocol/allocation-optimizer/commit/470a72e64c4b0358eb264858d49081adf795349b))


### Features

* Add opt_mode option to the config ([0b4534a](https://github.com/graphprotocol/allocation-optimizer/commit/0b4534aae0f784a87eca1e162adfb0c3d1b01427))
* Add slower but more accurate "optimal" path ([1629de7](https://github.com/graphprotocol/allocation-optimizer/commit/1629de7f49cec83a117af853e626cbb8e77e942b))
* Add warning for using optimal path ([0ced47f](https://github.com/graphprotocol/allocation-optimizer/commit/0ced47f3b3a10741073d845924cf7727ab3f3f62))

src/configuration.jl
# Copyright 2022-, The Graph Foundation
# SPDX-License-Identifier: MIT

"""
    configuredefaults!(config::AbstractDict)

Set default values for the config dictionary if the value was not specified in the config file.

# Config Specification
- `id::String`: The ID of the indexer for whom we're optimising. No default value.
- `network_subgraph_endpoint::String`: The network subgraph endpoint to query.
    By default, `"https://api.thegraph.com/subgraphs/name/graphprotocol/graph-network-mainnet"`
- `writedir::String`: The directory to which to write the results of optimisation.
    If don't specify `readdir`, `writedir` also specifies the path to which to save
    the input data tables. By default, `"."`
- `readdir::Union{String, Nothing}`: The directory from which to read saved data tables.
    This speeds up the process as we won't have to query the network subgraph for the
    relevant data. If you don't specify `readdir`, we will query your specified
    `network_subgraph_endpoint` for the data and write it to CSV files in `writedir`.
    This way, you can use your previous `writedir` as your `readdir` in future runs.
    By default, `nothing`
- `whitelist::Vector{String}`: A list of subgraph IPFS hashes that you want to consider
    as candidates to which to allocate. If you leave this empty, we'll assume all subgraphs
    are in the whitelist. By default, `String[]`
- `blacklist::Vector{String}`: A list of subgraph IPFS hashes that you do not want to
    consider allocating to. For example, this list could include broken subgraphs or
    subgraphs that you don't want to index. By default, `String[]`
- `frozenlist::Vector{String}`: If you have open allocations that you don't want to change,
    add the corresponding subgraph IPFS hashes to this list. By default, `String[]`
- `pinnedlist::Vector{String}`: If you have subgraphs that you absolutely want to be
    allocated to, even if only with a negligible amount of GRT, add it to this list.
    By default, `String[]`
- `allocation_lifetime::Integer`: The number of epochs for which you expect the allocations
    the optimiser finds to be open. By default, `28`
- `gas::Real`: The estimated gas cost in GRT to open/close allocations. By default, `100`
- `min_signal::Real`: The minimum amount of signal in GRT that must be on a subgraph
    in order for you to consider allocating to it. By default, `100`
- `max_allocations::Integer`: The maximum number of new allocations you'd like the optimiser
    to consider opening. By default, `10`
- `num_reported_options::Integer`: The number of proposed allocation strategies to report.
    For example, if you select `10` we'd report best 10 allocation strategies ranked by
    profit. By default, `1`
- `verbose::Bool`: If true, the optimiser will print details about what it is doing to
    stdout. By default, `false`
- `execution_mode::String`: How the optimiser should execute the allocation strategies it
    finds. Options are `"none"`, which won't do anything, `"actionqueue"`, which will
    push actions to the action queue, and `"rules"`, which will generate indexing rules.
    By default, `"none"`
- `indexer_url::Union{String, Nothing}`: The URL of the indexer management server you want
    to execute the allocation strategies on. If you specify `"actionqueue"`, you must also
    specify `indexer_url`. By default, `nothing`
- `opt_mode::String`: We support three optimisation modes. One is `"fastnogas"`. This mode does
    not consider gas costs and optimises allocation amount over all subgraph deployments.
    Second one is `"fastgas"`. This mode considers gas, may not find the optimal strategy and
    could potentially fail to converge. This mode is also used to the top
    `num_reported_options` allocation strategies. The final mode is `"optimal"`.
    This mode is slower, but it satisfies stronger optimality conditions.
    It will find strategies at least as good as `"fastgas"`, but not guaranteed to be better.
    By default, `"optimal"`
    mode to find the optimal allocation. By default, `"optimal"`
- `protocol_network::String`: Defines the protocol network that allocation transactions
    should be sent to. The current protocol network options are "mainnet", "goerli",
    "arbitrum", and "arbitrum-goerli". By default, `"mainnet"`
- `syncing_networks::Vector{String}`: The list of syncing networks to support when selecting
    the set of possible subgraphs. This list should match the networks available to your
    graph-node. By default, the list is a singleton of your protocol network

```julia
julia> using AllocationOpt
julia> config = Dict{String, Any}("id" => "0xa")
julia> config = AllocationOpt.configuredefaults!(config)
Dict{String, Any} with 16 entries:
  "execution_mode"            => "none"
  "readdir"                   => nothing
  "writedir"                  => "."
  "num_reported_options"      => 1
  "id"                        => "0xa"
  "pinnedlist"                => String[]
  "indexer_url"               => nothing
  "gas"                       => 100
  "allocation_lifetime"       => 28
  "blacklist"                 => String[]
  "verbose"                   => false
  "min_signal"                => 100
  "network_subgraph_endpoint" => "https://api.thegraph.com/subgraphs/name/graphprotocol/graph-network-mainnet"
  "whitelist"                 => String[]
  "protocol_network"          => "mainnet"
  "syncing_networks"          => String["mainnet"]
```
"""
function configuredefaults!(config::AbstractDict)
    @assert haskey(config, "id")
    setdefault!(
        config,
        "network_subgraph_endpoint",
        "https://api.thegraph.com/subgraphs/name/graphprotocol/graph-network-mainnet",
    )
    setdefault!(config, "writedir", ".")
    setdefault!(config, "readdir", nothing)
    setdefault!(config, "whitelist", String[])
    setdefault!(config, "blacklist", String[])
    setdefault!(config, "frozenlist", String[])
    setdefault!(config, "pinnedlist", String[])
    setdefault!(config, "allocation_lifetime", 28)
    setdefault!(config, "gas", 100)
    setdefault!(config, "min_signal", 100)
    setdefault!(config, "max_allocations", 10)
    setdefault!(config, "num_reported_options", 1)
    setdefault!(config, "indexer_url", nothing)
    setdefault!(config, "execution_mode", "none")
    setdefault!(config, "verbose", false)
    setdefault!(config, "opt_mode", "optimal")
    setdefault!(config, "protocol_network", "mainnet")
    setdefault!(config, "syncing_networks", String[config["protocol_network"]])
    return config
end

"""
    formatconfig!(config::AbstractDict)

Given a `config`, reformat values that need to be standardised.

```julia
julia> using AllocationOpt
julia> config = Dict("id" => "0xA")
julia> AllocationOpt.formatconfig!(config)
Dict{String, String} with 1 entry:
  "id" => "0xa"
```
"""
function formatconfig!(config::AbstractDict)
    config["id"] = config["id"] |> lowercase
    return config
end

"""
    readconfig(p::AbstractString)

Read the config file from path `p`. The config file must be specifed as a TOML.

See [`configuredefaults!`](@ref) to see which fields you should specify in the config.

```julia
julia> using AllocationOpt
julia> path = "myconfig.toml"
julia> config = AllocationOpt.readconfig(path)
Dict{String, Any} with 13 entries:
  "execution_mode"       => "none"
  "writedir"             => "data"
  "num_reported_options" => 2
  "id"                   => "0xd75c4dbcb215a6cf9097cfbcc70aab2596b96a9c"
  "pinnedlist"           => Union{}[]
  "gas"                  => 100
  "allocation_lifetime"  => 28
  "blacklist"            => Union{}[]
  "verbose"              => true
  "min_signal"           => 100
  "whitelist"            => Union{}[]
  "max_allocations"      => 5
  "frozenlist"           => Union{}[]
  "protocol_network"     => "mainnet"
  "syncing_networks"     => String["mainnet", "gnosis"]
```
"""
readconfig(p::AbstractString) = p |> TOML.parsefile
